fraud/bad_encoding.go

```diff
@@ -0,0 +1,179 @@
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
+			if index >= len(errShares)/2 || int(position) >= len(errShares)/2 {
+				namespaceID = consts.ParitySharesNamespaceID
+			}
+
+			shares[index] = &Share{namespaceID, share, &proof}
+		}
+	}
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares,
+		isRow:       isRow,
+		Position:    position,
+	}, nil
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to []byte
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEnconding{
+		Height:   p.BlockHeight,
+		Shares:   shares,
+		Position: uint32(p.Position),
+		IsRow:    p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
+	merkleRowRoots := header.DAH.RowsRoots
+	merkleColRoots := header.DAH.ColumnRoots
+	if int(p.Position) >= len(merkleRowRoots) || int(p.Position) >= len(merkleColRoots) {
+		return errors.New("invalid fraud proof: incorrect position of bad row/col")
+	}
+	if len(merkleRowRoots) != len(merkleColRoots) {
+		return errors.New("invalid fraud proof: invalid extended header")
+	}
+	if len(merkleRowRoots) != len(p.Shares) || len(merkleColRoots) != len(p.Shares) {
+		return errors.New("invalid fraud proof: invalid shares")
+	}
+
+	roots := merkleRowRoots
+	if p.isRow {
+		roots = merkleColRoots
+	}
+
+	shares := make([][]byte, len(roots))
+
+	// verify that Merkle proofs correspond to particular shares
+	for index, share := range p.Shares {
+		if share == nil {
+			continue
+		}
+		shares[index] = share.Raw
+		if ok := share.Validate(roots[index]); !ok {
+			return fmt.Errorf("invalid fraud proof: incorrect share received at position %d", index)
+		}
+	}
+
+	// rebuild a row/col
+	codec := rsmt2d.NewRSGF8Codec()
+	rebuiltShares, err := codec.Decode(shares)
+	if err != nil {
+		return err
+	}
+	rebuiltExtendedShares, err := codec.Encode(rebuiltShares)
+	if err != nil {
+		return err
+	}
+	rebuiltShares = append(rebuiltShares, rebuiltExtendedShares...)
+
+	tree := NewErasuredNamespacedMerkleTree(uint64(len(shares) / 2))
+	for i, share := range rebuiltShares {
+		tree.Push(share, rsmt2d.SquareIndex{Axis: uint(p.Position), Cell: uint(i)})
+	}
+
+	merkleRoot := merkleColRoots[p.Position]
+	if p.isRow {
+		merkleRoot = merkleRowRoots[p.Position]
+	}
+
+	// compare Merkle Roots
+	if bytes.Equal(tree.Root(), merkleRoot) {
+		return errors.New("invalid fraud proof: merkle root matches")
+	}
+
+	return nil
+}
+
+func buildTreeFromDataRoot(tree *ErasuredNamespacedMerkleTree, root [][]byte, index int) error {
+	for idx, data := range root {
```

Comment:
Similar to [above](https://github.com/celestiaorg/celestia-node/pull/545/files#r843229315): why is the row/Col leaf data called root here?

Reply:
Renamed to `buildTreeFromLeaves`